site/content/blog/2020-06-10-opentracing.md

```diff
@@ -0,0 +1,576 @@
+
+First of all, we kick of AST parser on our input file that contains the interface and start walking through the found nodes:
+
+```go
+	fset := token.NewFileSet() // positions are relative to fset
+
+	file, err := os.Open("animal.go")
+	if err != nil {
+		return nil, fmt.Errorf("Unable to open %s file: %w", inputFile, err)
+	}
+	src, err := ioutil.ReadAll(file)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+	f, err := parser.ParseFile(fset, "animal.go", src, parser.AllErrors|parser.ParseComments)
+	if err != nil {
+		return nil, err
+	}
+
+	ast.Inspect(f, func(n ast.Node) bool {
+		// ... handle the found nodes
+	})
+```
+
+To differentiate interface methods from other AST nodes, we can do the following: 
+
+```go
+	ast.Inspect(f, func(n ast.Node) bool {
+		switch x := n.(type) {
+		case *ast.TypeSpec:
+			if x.Name.Name == "IAnimal" {
+				for _, method := range x.Type.(*ast.InterfaceType).Methods.List {
+					methodName := method.Names[0].Name
+					// here we can parse all the required information about the method
+					methods[methodName] = extractMethodMetadata(method, src)
+				}
+			}
+		}
+		return true
+	})
+```
+
+Let's define a couple of struct to help us collect information about methods:
+
+```go
+type methodParam struct {
+	Name string
+	Type string
+}
+
+type methodData struct {
+	Params        []methodParam
+	Results       []string
+}
+
+
+// for each found method, we'll store it's name, params with their types and return types in methods := map[string]methodData {}
+```
+
+Now let's write a short function to populate these structs with metadata about a method:
+
+```go
+func formatNode(src []byte, node ast.Expr) string {
+	return string(src[node.Pos()-1 : node.End()-1])
+}
+
+func extractMethodMetadata(method *ast.Field, src []byte) methodData {
+	params := []methodParam{}
+	results := []string{}
+	e := method.Type.(*ast.FuncType)
+	if e.Params != nil {
+		for _, param := range e.Params.List {
+			for _, paramName := range param.Names {
+				paramType := (formatNode(src, param.Type))
+				params = append(params, methodParam{Name: paramName.Name, Type: paramType})
+			}
+		}
+	}
+
+	if e.Results != nil {
+		for _, r := range e.Results.List {
+			typeStr := formatNode(src, r.Type)
+			if len(r.Names) > 0 {
+				for _, k := range r.Names {
+					results = append(results, fmt.Sprintf("%s %s", k.Name, typeStr))
+				}
+			} else {
+				results = append(results, typeStr)
+			}
+		}
+	}
+	return methodData{Params: params, Results: results}
+}
+```
+
+Now we can run the parser on our interface and we'll get something like: `map[Speak:{Params:[{Name:x Type:int}] Results:[]}]`
+As you can see, we collected all the information needed about interface methods and we can now move on to generating decorator with this data.
```

Comment:
```suggestion
As you can see, we collected all the information needed about interface methods and we can now move on to generating a decorator with this data.
```
^^ Not sure if using "a" (or "the") is correct in this case? Seems like it should be but let me know!

Reply:
should be 'the' i think.